pca_ica/extract_cells.m

```diff
@@ -0,0 +1,353 @@
+    U = gather(U);
+    one_norms = gather(one_norms);
+    norms_U = gather(norms_U);
+    
+else % CPU
+    avail_mem = compute_available_memory(0);
+    f = avail_mem/4 ; % maximum available element size
+    f = f*mem_occup_scale_CPU;
+    chunk_size = floor(f/ (2*N+num_pcs)); %# of pixels at once
+    
+    one_norms = zeros(1,N);
+    for i = 1:chunk_size:N
+        fin = min(chunk_size-1,N-i);
+        Q = U_norm(i:i+fin,:)';
+        one_norms(i:i+fin) = sum(abs(U*Q),1);
+        clear Q;
+    end
+
+end
+
+% Extract filters recursively using the 1-norm criterion
+idx_possible = 1:N;
+F = zeros(N,max_num);
+inv_quality = zeros(1,max_num);
+acc=0;
+
+while true
+    acc = acc+1;
+    [val,idx_this] = min(one_norms(idx_possible));
+    idx_this = idx_possible(idx_this);
+    sig_this = U*U_norm(idx_this,:)';    
+    correl = sig_this.*(1./norms_U);
+    idx_possible = intersect(find(abs(correl)<corr_thresh),idx_possible);
+    F(:,acc) = sig_this;
+    inv_quality(:,acc) = val;
+    % Termination condition
+    if isempty(idx_possible) || acc==max_num
+        break;
+    end
+    
+end
+
+% Truncate F in case there are less components than max_num
+F = F(:,1:acc);
+inv_quality = inv_quality(1:acc)/sqrt(N);
+
+if is_trimmed % Untrim the pixels
+    F_temp = zeros(h*w,acc);
+    F_temp(idx_kept,:) = F;
+    F = F_temp;
+end
+
+fprintf('\t \t \t Extracted %d potential cells \n',acc);
+fprintf('%s: Cleaning filters and removing duplicates...\n',datestr(now));
+[F,idx] = modify_filters(F,filter_thresh);
+inv_quality = inv_quality(idx);
+
+figure,
+plot(inv_quality,'LineWidth',1.5);
+title('Inverse quality metric for the extracted potential cells','Fontsize',16);
+xlabel('Cell index','Fontsize',16)
+ylabel('Inverse quality(normalized to 1)','Fontsize',16)
+
+fprintf('%s: Loading movie for trace extraction...\n',datestr(now));
+M = load_movie(movie_source);
+
+% Extract time traces
+fprintf('%s: Extracting traces...\n',datestr(now));
+M = reshape(M,h*w,t);
+idx_nonzero = find(sum(F,2)>0);
+M_small = M(idx_nonzero,:);
+F_small = F(idx_nonzero,:);
+T = (F_small'*F_small+0*eye(size(F,2))) \  (F_small'*M_small);
+clear M;
+
+% Reconstruction settings
+info.type = 'PCA+SSS';
+info.num_pairs = size(F,2);
+
+filters = reshape(F,h,w,size(F,2));
+traces = T';
+
+% Remove baseline from the traces
+if do_baseline_fix
+    for k = 1:size(F,2);
+        traces(:,k) = fix_baseline(traces(:,k));
+    end
+end
+
+% Save the result to mat file
+%------------------------------------------------------------
+timestamp = datestr(now, 'yymmdd-HHMMSS');
+rec_savename = sprintf('rec_%s.mat', timestamp);
+
+fprintf('%s: Writing the output to file... \n',datestr(now));
+fprintf('%s: Output will be saved in %s \n',datestr(now),rec_savename);
+
+save(rec_savename, 'info', 'filters', 'traces');
+
+fprintf('%s: All done! \n',datestr(now));
+
+
+
+%-------------------
+% Internal functions
+%-------------------
+
+function avail_mem = compute_available_memory(is_GPU)
```

Comment:
Just a comment, I think it would have been cleaner to write two functions: `function gpu_mem = compute_gpu_memory()` and `function cpu_mem = compute_cpu_memory()` here, instead of using a `is_GPU` flag.